File: arrow.js
#!/usr/bin/env node

// print cli usage
function printUsage() {
  console.log("Usage: node arrow.js x\n" +
              " E.g.: node arrow.js 4\n" +
              " Note: x have to be >= 2 && <=20 \n" +
              "\n     let's try again (: \n")
}

// read cli input
const args = process.argv
var   x = ~~args[2]
if ( (typeof x != "number") || (x <= 2) || (x >= 20) ) {
  printUsage();
} else {
  drawArrow(x)
}

// draw the arrow
function drawArrow(x) {
  var col = (x*2)-1      // column
  var row = x*2          // row
  var   m = 0;          // mirror counter for middle of the head

  //console.log("x="+x+"; col="+col+"; row="+row+";")
  console.log("x="+x+";")

  // for each rows
  for (i=0;i<row;i++) {
    line = ''
    // draw all the head but the first row (same as tail)
    if (i < x && i!=0) {
      // last row of the head
      if (i == x-1) {
        Array(row-1).fill(1).forEach(function() { line += '*' })
      } else {
        m++
        for (l=1; l<=col; l++) {
          // use the mirror counter
          if (l == x - m || l == x + m) {
            line += '*'
          } else {
            line += ' '
          }
        }
      }
    // draw the tail and the first line
    } else {
      for (l=1; l<=col; l++) {
        // draw the * in the middle
        if (l == x) {
          line += '*'
        } else {
          line += ' '
        }
      }
    }
    console.log(line)
  }
}

File: arrow.dart
#!/usr/bin/env dart
// 2018-06-29 EPFL DOJO — https://www.dartlang.org/

void main(List<String> args) {

  // Dart 2: int length = int.tryParse(args[0]) ?? 8;
  var length = int.parse(args[0], onError: (source) => null);
  assert(length is int);

  // Draw the arrow
  print(tip(length));
  head(length);
  for (final t in tail(length)) {
    print(t);
  }

}

string tip(length) {
  return " " * length + "*";
}

string head(length) {
  for (var i = 1; i < length; i++) {
    print(" " * (length-i) + '*' + " " * (2*i-1) + '*') ;
  }
  print('* ' * length + '*');
}

List tail(length) {
  var list = new List(length);
  for(var i = 0; i < length; i++) {
    list[i] = tip(length);
  }
  return list;
}
